sim/shaman/item_sets_phase_4.go

```diff
@@ -0,0 +1,307 @@
+	Bonuses: map[int32]core.ApplyEffect{
+		// +40 Attack Power, up to 23 increased damage from spells, and up to 44 increased healing from spells.
+		2: func(agent core.Agent) {
+			c := agent.GetCharacter()
+			c.AddStats(stats.Stats{
+				stats.AttackPower:       40,
+				stats.RangedAttackPower: 40,
+				stats.SpellDamage:       23,
+				stats.HealingPower:      44,
+			})
+		},
+		// 6% chance on mainhand autoattack and 4% chance on spellcast to increase your damage and healing done by magical spells and effects by up to 95 for 10 sec.
+		4: func(agent core.Agent) {
+			c := agent.GetCharacter()
+
+			procAura := c.NewTemporaryStatsAura("The Furious Storm", core.ActionID{SpellID: 27775}, stats.Stats{stats.SpellPower: 95}, time.Second*10)
+			handler := func(sim *core.Simulation, spell *core.Spell, _ *core.SpellResult) {
+				procAura.Activate(sim)
+			}
+
+			core.MakeProcTriggerAura(&c.Unit, core.ProcTrigger{
+				Name:       "Item - The Furious Storm Proc (MH Auto)",
+				Callback:   core.CallbackOnSpellHitDealt,
+				Outcome:    core.OutcomeLanded,
+				ProcMask:   core.ProcMaskMeleeMHAuto,
+				ProcChance: 0.06,
+				Handler:    handler,
+			})
+			core.MakeProcTriggerAura(&c.Unit, core.ProcTrigger{
+				Name:       "Item - The Furious Storm Proc (Spell Cast)",
+				Callback:   core.CallbackOnCastComplete,
+				ProcMask:   core.ProcMaskSpellDamage | core.ProcMaskSpellHealing,
+				ProcChance: 0.04,
+				Handler:    handler,
+			})
+		},
+		// +8 All Resistances.
+		6: func(agent core.Agent) {
+			c := agent.GetCharacter()
+			c.AddResistances(8)
+		},
+		// +200 Armor.
+		8: func(agent core.Agent) {
+			c := agent.GetCharacter()
+			c.AddStat(stats.Armor, 200)
+		},
+	},
+})
+
+var ItemSetEarthfuryEruption = core.NewItemSet(core.ItemSet{
+	Name: "Earthfury Eruption",
+	Bonuses: map[int32]core.ApplyEffect{
+		// The radius of your totems that affect friendly targets is increased to 40 yd.
+		2: func(agent core.Agent) {
+			// Nothing to do
+		},
+		4: func(agent core.Agent) {
+			shaman := agent.(ShamanAgent).GetShaman()
+			shaman.applyT1Elemental4PBonus()
+		},
+		6: func(agent core.Agent) {
+			shaman := agent.(ShamanAgent).GetShaman()
+			shaman.applyT1Elemental6PBonus()
+		},
+	},
+})
+
+// Your Lightning Bolt critical strikes have a 35% chance to reset the cooldown on Lava Burst and Chain Lightning and make the next Lava Burst, Chain Heal, or Chain Lightning within 10 sec instant.
+func (shaman *Shaman) applyT1Elemental4PBonus() {
+	label := "S03 - Item - T1 - Shaman - Elemental 4P Bonus"
+	if shaman.HasAura(label) {
+		return
+	}
+
+	core.MakePermanent(shaman.RegisterAura(core.Aura{
+		Label: label,
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell.SpellCode == SpellCode_ShamanLightningBolt && spell.ProcMask.Matches(core.ProcMaskSpellDamage) && result.DidCrit() && sim.Proc(.35, "Power Surge") {
+				shaman.PowerSurgeDamageAura.Activate(sim)
+			}
+		},
+	}))
+}
```

Comment:
Bonus defined as follows:
1. Comment with the spell tooltip description
2. Class method `apply[Tier#][SpecName][#Pieces]Bonus()`
3. Define the label then return if the character already has the aura registered. This is more important for more complex bonuses where `GetOrRegisterAura` may not necessarily prevent other side effects and is a more reliable pattern we can follow